File: windows/VoiceKey.Core/DictationMode.cs
namespace VoiceKey.Core;

/// <summary>
/// Per-app dictation mode ("Power Mode"): the foreground app decides how a
/// transcript is punctuated before it is inserted at the cursor.
/// </summary>
public enum DictationMode
{
    /// <summary>Terminals, editors, IDEs — spoken commands should not gain a trailing period.</summary>
    Code,
    /// <summary>Mail clients — sentences should end in punctuation.</summary>
    Email,
    /// <summary>Everything else — the transcript is inserted verbatim.</summary>
    Standard,
}

public static class DictationModes
{
    /// <summary>
    /// Whether a finished transcript should be typed at the cursor at all.
    ///
    /// It should not when VoiceKey itself is frontmost: the cursor is then in
    /// VoiceKey's own window, and pasting there types into the search field —
    /// which filters the library down to the transcript just spoken. The
    /// transcript is in the list either way, so the clipboard is the honest
    /// place for it. An unknown app on either side is assumed to be elsewhere.
    /// </summary>
    public static bool InsertsAtCursor(string? frontmostProcess, string? ownProcess)
    {
        if (frontmostProcess is null || ownProcess is null) return true;
        return !string.Equals(frontmostProcess, ownProcess, StringComparison.OrdinalIgnoreCase);
    }

    /// <summary>
    /// Executable names, matched case-insensitively without their <c>.exe</c> suffix.
    /// macOS keys this table on bundle IDs; Windows has no equivalent stable identifier,
    /// so the executable name stands in.
    /// </summary>
    private static readonly HashSet<string> CodeProcesses = new(StringComparer.OrdinalIgnoreCase)
    {
        // terminals
        "windowsterminal", "wt", "cmd", "powershell", "pwsh",
        "alacritty", "wezterm-gui", "conemu64", "hyper",
        // editors and IDEs
        "code", "code - insiders", "cursor", "devenv", "zed",
        "sublime_text", "notepad++", "githubdesktop", "windsurf",
        // JetBrains ships one executable per IDE; macOS matches its bundle IDs by
        // prefix, but "idea"/"pycharm" are too short to prefix-match safely here.
        "idea64", "pycharm64", "webstorm64", "goland64", "rider64",
        "clion64", "phpstorm64", "rubymine64", "datagrip64", "rustrover64",
    };

    private static readonly HashSet<string> EmailProcesses = new(StringComparer.OrdinalIgnoreCase)
    {
        "outlook", "olk", "hxoutlook", "thunderbird", "mailclient",
    };

    /// <summary>
    /// Mode for the foreground app. Accepts a bare process name, an executable name, or a
    /// full module path — null or unknown maps to <see cref="DictationMode.Standard"/>.
    /// </summary>
    public static DictationMode ForProcess(string? processName)
    {
        var name = Normalised(processName);
        if (name.Length == 0) return DictationMode.Standard;
        if (CodeProcesses.Contains(name)) return DictationMode.Code;
        if (EmailProcesses.Contains(name)) return DictationMode.Email;
        return DictationMode.Standard;
    }

    private static string Normalised(string? processName)
    {
        if (string.IsNullOrWhiteSpace(processName)) return "";
        var name = Path.GetFileName(processName.Trim());
        return name.EndsWith(".exe", StringComparison.OrdinalIgnoreCase)
            ? name[..^4]
            : name;
    }

    /// <summary>Mode-specific transcript formatting. Pure.</summary>
    public static string Format(this DictationMode mode, string text) => mode switch
    {
        DictationMode.Code => FormattedForCode(text),
        DictationMode.Email => FormattedForEmail(text),
        _ => text,
    };

    /// <summary>
    /// Drops exactly one trailing period — but leaves "…" and "..." intact, since
    /// those are dictated content rather than sentence punctuation.
    /// </summary>
    private static string FormattedForCode(string text)
    {
        var trimmed = text.TrimEnd();
        if (!trimmed.EndsWith('.')) return trimmed;
        var withoutPeriod = trimmed[..^1];
        return withoutPeriod.EndsWith('.') ? trimmed : withoutPeriod;
    }

    /// <summary>
    /// Appends a period when the sentence trails off on a letter or digit;
    /// anything already ending in punctuation (. ! ? : …) is left alone.
    /// </summary>
    private static string FormattedForEmail(string text)
    {
        var trimmed = text.TrimEnd();
        if (trimmed.Length == 0) return trimmed;
        var last = trimmed[^1];
        return char.IsLetter(last) || char.IsNumber(last) ? trimmed + "." : trimmed;
    }
}

File: macos/Sources/VoiceKeyCore/DictationMode.swift
import Foundation

/// Per-app dictation mode ("Power Mode"): the frontmost app decides how a
/// transcript is punctuated before it is inserted at the cursor.
public enum DictationMode: Equatable {
    /// Terminals, editors, IDEs — spoken commands should not gain a trailing period.
    case code
    /// Mail clients — sentences should end in punctuation.
    case email
    /// Everything else — the transcript is inserted verbatim.
    case standard

    /// Bundle IDs are compared lower-cased, so every entry here is lower-cased too.
    private static let codeBundleIDs: Set<String> = [
        "com.apple.terminal",
        "com.googlecode.iterm2",
        "com.mitchellh.ghostty",
        "com.microsoft.vscode",
        "com.todesktop.230313mzl4w4u92",
        "com.apple.dt.xcode",
        "dev.zed.zed",
        "com.sublimetext.4",
        "com.github.githubclient",
    ]

    /// JetBrains ships one bundle ID per IDE (idea, pycharm, goland, …), so they are matched by prefix.
    private static let codeBundleIDPrefixes: [String] = [
        "com.jetbrains.",
    ]

    private static let emailBundleIDs: Set<String> = [
        "com.apple.mail",
        "com.microsoft.outlook",
        "com.readdle.sparkdesktop",
        "com.superhuman.electron",
    ]

    /// Mode for the frontmost app. nil bundle ID → .standard.
    public static func mode(forBundleID bundleID: String?) -> DictationMode {
        guard let identifier = bundleID?.lowercased() else { return .standard }
        if codeBundleIDs.contains(identifier) { return .code }
        if codeBundleIDPrefixes.contains(where: identifier.hasPrefix) { return .code }
        if emailBundleIDs.contains(identifier) { return .email }
        return .standard
    }

    /// Whether a finished dictation should be typed at the cursor at all.
    ///
    /// VoiceKey never types into itself. Dictating with its own window in front
    /// pastes into whatever there has focus — the library's search field — and
    /// the list then filters down to the transcript just spoken, which reads as
    /// every earlier one having been lost. The transcript is recorded and copied
    /// instead; the row's own "Insert" button is the way back out, and it steps
    /// the app aside first.
    public static func insertsAtCursor(frontmostBundleID: String?, ownBundleID: String?) -> Bool {
        guard let own = ownBundleID?.lowercased(),
              let frontmost = frontmostBundleID?.lowercased()
        else { return true }
        return frontmost != own
    }

    /// Mode-specific transcript formatting. Pure.
    public func format(_ text: String) -> String {
        switch self {
        case .code: return Self.formattedForCode(text)
        case .email: return Self.formattedForEmail(text)
        case .standard: return text
        }
    }

    /// Drops exactly one trailing period — but leaves "…" and "..." intact, since
    /// those are dictated content rather than sentence punctuation.
    private static func formattedForCode(_ text: String) -> String {
        let trimmed = trimmingTrailingWhitespace(text)
        guard trimmed.hasSuffix(".") else { return trimmed }
        let withoutPeriod = String(trimmed.dropLast())
        guard !withoutPeriod.hasSuffix(".") else { return trimmed }
        return withoutPeriod
    }

    /// Appends a period when the sentence trails off on a letter or digit;
    /// anything already ending in punctuation (. ! ? : …) is left alone.
    private static func formattedForEmail(_ text: String) -> String {
        let trimmed = trimmingTrailingWhitespace(text)
        guard let last = trimmed.last, last.isLetter || last.isNumber else { return trimmed }
        return trimmed + "."
    }

    private static func trimmingTrailingWhitespace(_ text: String) -> String {
        String(text.reversed().drop(while: { $0.isWhitespace }).reversed())
    }
}
